src/profile/ProfileForm.tsx

```diff
@@ -0,0 +1,494 @@
+import axios from "axios";
+import React, { forwardRef, useImperativeHandle, useReducer, useRef, useState } from "react";
+import { Button, Form, Row, Col, Alert, OverlayTrigger, Popover, InputGroup } from "react-bootstrap";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faEye, faEyeSlash } from "@fortawesome/free-solid-svg-icons";
+import nanoid from "nanoid";
+
+import { User, UserAPIError } from "../shared/types"
+import { useAuth } from "../auth/AuthProvider";
+import ERRORS from "../shared/errors";
+import { getApiUrl } from "../api/api";
+
+interface ChangedUser {
+  username?: string;
+  name?: string;
+  first_name?: string;
+  last_name?: string;
+  email?: string;
+  is_editor?: boolean;
+  writer_name?: string;
+  cur_password?: string;
+  password?: string;
+}
+
+interface ProfileState {
+  errors: UserAPIError;
+  generalErrors: string[];
+  successMessage: string;
+  isLoading: boolean;
+  isChecking: boolean;
+  changedUser: ChangedUser;
+}
+
+type ProfileAction =
+  { type: 'is loading' } |
+  { type: 'is checking' } |
+  { type: 'done checking' } |
+  { type: 'done loading success' } |
+  { type: 'done loading error', errors: UserAPIError | string | string[] } |
+  { type: 'set error', errors: UserAPIError } |
+  { type: 'set general error', errors: string[] } |
+  { type: 'set success message', message: string } |
+  { type: 'set data', data: ChangedUser, errors?: UserAPIError };
+
+const profileReducer = (state: ProfileState, action: ProfileAction): ProfileState => {
```

Comment:
I anticipate us having a lot of forms, we should use Formik instead of handrolling a reducer for every single one.